templates/commands/render.go

```diff
@@ -235,24 +275,237 @@ func (r *Render) realRun(ctx context.Context, rp *runParams) (outErr error) {
+	// short writes always return error.
+	if _, err := sp.stdout.Write([]byte(msg)); err != nil {
+		return fmt.Errorf("error writing to stdout: %w", err)
+	}
+
+	return nil
+}
+
+func actionInclude(ctx context.Context, i *model.Include, sp *stepParams) error {
+	for _, p := range i.Paths {
+		// Paths may contain template expressions, so render them first.
+		walkRelPath, err := parseAndExecuteGoTmpl(p, sp.inputs)
+		if err != nil {
+			return errWithPos(p.Pos, `error compiling as go-template: %w`, err)
+		}
+
+		if err := safeRelPath(p.Val); err != nil {
+			return errWithPos(p.Pos, "invalid path: %w", err)
+		}
+
+		absSrc := filepath.Join(sp.templateDir, walkRelPath)
+		absDst := filepath.Join(sp.scratchDir, walkRelPath)
+
+		if _, err := sp.fs.Stat(absSrc); err != nil {
+			if errors.Is(err, os.ErrNotExist) {
+				return errWithPos(p.Pos, "include path doesn't exist: %q", walkRelPath)
+			}
+		}
+
+		if err := copyRecursive(p.Pos, absSrc, absDst, sp.fs); err != nil {
+			return errWithPos(p.Pos, "copying failed: %w", err)
+		}
+	}
+	return nil
+}
+
+// "from" may be a file or directory. "pos" is only used for error messages.
+func copyRecursive(pos *model.ConfigPos, from, to string, rfs renderFS) error {
+	return fs.WalkDir(rfs, from, func(path string, d fs.DirEntry, err error) error {
+		if err != nil {
+			return err // There was some filesystem error. Give up.
+		}
+
+		// We don't have to worry about symlinks here because we passed
+		// DisableSymlinks=true to go-getter.
+
+		relToSrc, err := filepath.Rel(from, path)
+		if err != nil {
+			return errWithPos(pos, "filepath.Rel(%s,%s): %w", from, path, err)
+		}
+
+		dst := filepath.Join(to, relToSrc)
+
+		// The spec file may specify a file to copy that's deep in a
+		// directory tree, without naming its parent directory. We can't
+		// rely on WalkDir having traversed the parent directory of $path,
+		// so we must create the target directory if it doesn't exist.
+		dirToCreate := dst
+		if !d.IsDir() {
+			dirToCreate = filepath.Dir(dst)
+		}
+		if err := rfs.MkdirAll(dirToCreate, 0o700); err != nil {
+			return errWithPos(pos, "MkdirAll(): %w", err)
+		}
+		if d.IsDir() {
+			return nil
+		}
+
+		buf, err := rfs.ReadFile(path)
+		if err != nil {
+			return errWithPos(pos, "ReadFile(): %w", err)
+		}
+
+		// The permission bits for normal files in the scratch directory will be
+		// set conservatively to either 600 (u=rw) or 700 (u=rwx), depending on
+		// the owner-executable permission bit of the source file. These
+		// permissions will be further limited by the user's umask (see `man
+		// umask` if unfamiliar), so the actual permissions on disk might be
+		// more restrictive than we ask for.
+		info, err := rfs.Stat(path)
+		if err != nil {
+			return fmt.Errorf("Stat(): %w", err)
+		}
+		ownerExecuteBit := info.Mode() & 0o100
+		dstPerms := 0o600 | ownerExecuteBit
+
+		if err := rfs.WriteFile(dst, buf, dstPerms); err != nil {
```

Comment:
It can be a follow-up, but I think you're going to want an atomic write here. I've done this before, it's not fun: https://github.com/hashicorp/consul-template/blob/e97a929318455f1ff0d11360a746412d633a004c/renderer/renderer.go#L117-L214

Reply:
SGTM, I think this can wait. Added a TODO.